python/rmm/_lib/device_buffer.pyx

```diff
@@ -278,17 +278,35 @@ cdef class DeviceBuffer:
 
         return b
 
-    cdef size_t c_size(self):
-        return self.c_obj.get()[0].size()
+    cdef size_t c_size(self) except *:
+        if self.c_obj:
+            return self.c_obj.get()[0].size()
+        else:
+            return 0
 
-    cpdef void resize(self, size_t new_size):
+    cpdef void resize(self, size_t new_size) except *:
         self.c_obj.get()[0].resize(new_size)
 
-    cpdef size_t capacity(self):
-        return self.c_obj.get()[0].capacity()
+    cpdef size_t capacity(self) except *:
+        if self.c_obj:
+            return self.c_obj.get()[0].capacity()
+        else:
+            return 0
 
-    cdef void* c_data(self):
-        return self.c_obj.get()[0].data()
+    cdef void* c_data(self) except *:
+        if self.c_obj:
+            return self.c_obj.get()[0].data()
+        else:
+            return <void*><uintptr_t>0
+
+    cdef device_buffer c_release(self) except *:
+        """
+        Releases ownership the data held by this DeviceBuffer.
+        """
+        if self.c_obj:
+            return cython.operator.dereference(self.c_obj.release())
+        else:
+            return device_buffer(0)
```

Comment:
Could we assign and empty `device_buffer` to `c_obj` after we release the existing one? That should protect against errors in other cases.

Reply:
So we now call `move()` on the device buffer underlying `c_obj` as Jake suggested, and that has the same effect as assigning a unique_ptr to an empty device_buffer to `c_obj`  :)

Reply:
Ah ok. Didn't realize `move` would ensure this. Thought that was UB.

Reply:
Note that `move` by itself does _not_ guarantee this. It's 100% defined by the class's move constructor about what is the state of a "moved from" object. 

Here's what `std::vector` says: 

> Constructs the container with the contents of other using move semantics. Allocator is obtained by move-construction from the allocator belonging to other. After the move, other is guaranteed to be empty().

Here's what `device_buffer` says: https://github.com/rapidsai/rmm/blob/branch-0.15/include/rmm/device_buffer.hpp#L155